File: PlanovaKit/Sources/PlanovaKit/TripDiff.swift
import Foundation

/// A CloudKit-agnostic description of which records changed between two
/// snapshots of a trip. The sync layer turns these into CKRecord.IDs.
public enum TripRecordRef: Hashable, Sendable {
    case tripMeta
    case day(UUID)
    case item(UUID)
}

public enum TripDiff {

    public static func changes(old: Trip?, new: Trip)
        -> (saves: [TripRecordRef], deletes: [TripRecordRef]) {
        guard let old else {
            return ([.tripMeta] + new.days.map { .day($0.id) } + new.items.map { .item($0.id) }, [])
        }
        var saves: [TripRecordRef] = []
        var deletes: [TripRecordRef] = []

        let metaChanged = old.name != new.name || old.startDate != new.startDate
            || old.endDate != new.endDate || old.destinations != new.destinations
            || old.schemaVersion != new.schemaVersion
        if metaChanged { saves.append(.tripMeta) }

        let oldDays = Dictionary(uniqueKeysWithValues: old.days.map { ($0.id, $0) })
        let newDays = Dictionary(uniqueKeysWithValues: new.days.map { ($0.id, $0) })
        for (id, day) in newDays where oldDays[id] != day { saves.append(.day(id)) }
        for id in oldDays.keys where newDays[id] == nil { deletes.append(.day(id)) }

        let oldItems = Dictionary(uniqueKeysWithValues: old.items.map { ($0.id, $0) })
        let newItems = Dictionary(uniqueKeysWithValues: new.items.map { ($0.id, $0) })
        for (id, item) in newItems where oldItems[id] != item { saves.append(.item(id)) }
        for id in oldItems.keys where newItems[id] == nil { deletes.append(.item(id)) }

        return (saves, deletes)
    }
}

File: webapp/src/sync/tripDiff.ts
import type { Trip } from "../model/trip";
import type { EntityKind } from "./types";

export interface EntityRef { kind: EntityKind; id: string; }
export interface TripDiffResult { saves: EntityRef[]; deletes: EntityRef[]; }

const eq = (a: unknown, b: unknown): boolean => JSON.stringify(a) === JSON.stringify(b);

/** Mirror of the Swift TripDiff: which entities changed between snapshots. */
export function diffTrip(old: Trip | undefined, next: Trip): TripDiffResult {
  if (!old) {
    return {
      saves: [{ kind: "trip", id: next.id },
        ...next.days.map((d) => ({ kind: "day" as const, id: d.id })),
        ...next.items.map((i) => ({ kind: "item" as const, id: i.id }))],
      deletes: [],
    };
  }
  const saves: EntityRef[] = [];
  const deletes: EntityRef[] = [];

  const metaChanged = old.name !== next.name || old.startDate !== next.startDate
    || old.endDate !== next.endDate || !eq(old.destinations, next.destinations)
    || old.schemaVersion !== next.schemaVersion;
  if (metaChanged) saves.push({ kind: "trip", id: next.id });

  diffList(old.days, next.days, "day", saves, deletes);
  diffList(old.items, next.items, "item", saves, deletes);
  return { saves, deletes };
}

function diffList<T extends { id: string }>(
  oldArr: T[], newArr: T[], kind: EntityKind, saves: EntityRef[], deletes: EntityRef[],
): void {
  const oldMap = new Map(oldArr.map((x) => [x.id, x]));
  const newMap = new Map(newArr.map((x) => [x.id, x]));
  for (const [id, v] of newMap) if (!eq(oldMap.get(id), v)) saves.push({ kind, id });
  for (const id of oldMap.keys()) if (!newMap.has(id)) deletes.push({ kind, id });
}
